modules/dists/HashedDist.chpl

```diff
@@ -42,15 +47,77 @@ record DefaultMapper {
 //
 // The distribution class
 //
-class UserMapAssoc : BaseDist {
+
+/*
+
+The Hashed distribution stores an associative domain or array across many
+locales. Each index (or key) in the domain is stored in a locale based upon that
+index. When constructing a Hashed distribution, you can optionally provide
+provide a `mapper` function or function object to compute the destination
+locale for each index.
+
+The `mapper` provided can be a class, record, or first class function.  When
+called, it will be passed the index and the targetLocales array that the Hashed
+distribution was initialized with. For example, the following record
+declares a custom mapper:
+
+.. code-block:: chapel
+
+  record CustomMapper {
+    proc this(ind:string, targetLocs: [?D] locale) : D.idxType {
+      const numlocs = targetLocs.domain.size;
+      // use the first digit of the string to choose the destination locale
+      var byte: int = ascii(ind);
+      return byte % numlocs;
+    }
+  }
+
+If a custom mapper is not provided, a default mapper will be used. The default
```

Comment:
Is the default mapper the same as the local hash function that we use within a locale to map to the open addressing slots?  If so, I worry that this could result in weird interference effects in which since each locale is getting a specific fraction of the keys that the local hashing wouldn't populate the table evenly (I'm not sure I'm being clear).  Put another way, I wonder whether we should have two default hash functions for each type, one for inter-locale, one for intra-locale.  All that said, I wouldn't hold this PR up over this (if it's even true that we're using the same hash both times).

Reply:
Actually there's some interesting work on hashing that points out that if you have 2 completely different hash functions, the combinations of those are almost as good as a 3rd completely different hash function. So, it'd be appealing, if we are concerned about such things, to provide 2 hash functions, and then to use a different combination of these in the mapper than we use in the hashtable itself. Let me know if you think I should make an issue proposing that.

(I was thinking about proposing that our hash functions be "keyed" by an integer argument; in which case if providing hash functions for user types becomes user facing, they'd only need to create one overload, but could use the "key" to choose totally different functions if desired. Point being, asking all types to have "2 hash functions" doesn't necessarily mean that there are actually 2 functions to implement).

Reply:
I like the keyed hashing function concept.